torch_geometric/explain/base.py

```diff
@@ -0,0 +1,121 @@
+from abc import abstractmethod
+from typing import Optional
+
+import torch
+
+from torch_geometric.data import Data
+from torch_geometric.explain.explanations import Explanation
+
+
+class ExplainerAlgorithm(torch.nn.Module):
+    """Abstract class for explanation algorithms."""
+    @abstractmethod
+    def loss(self, y_hat: torch.Tensor, y: torch.Tensor) -> torch.Tensor:
+        r"""
+        This method compute the loss to be used for the explanation algorithm.
+        Subclasses should override this method to define their own loss.
+
+        Args:
+            y_hat (torch.Tensor): the output of the explanation algorithm.
+                (e.g. the forward pass of the model with the mask applied).
+            y (torch.Tensor): the reference output.
+        """
+
+    @torch.no_grad()
+    def get_initial_prediction(
+        self,
+        model: torch.nn.Module,
+        g: Data,
+        batch: Optional[torch.Tensor] = None,
+        **kwargs,
+    ):
+        """Return the initial prediction of the model.
+
+        Args:
+            model (torch.nn.Module): the model to explain.
+            g (torch_geometric.data.Data): the input graph.
+            batch (torch.Tensor, optional): batch indicator.
+        """
+        out = model(
+            g,
+            **dict(batch=batch, **kwargs),
+        )
+        return out
+
+    @abstractmethod
+    def forward(
+        self,
+        x: torch.Tensor,
+        edge_index: torch.Tensor,
+        *,
+        model: torch.nn.Module,
+        target: torch.Tensor,
+        target_index: int = 0,
+        batch: Optional[torch.Tensor] = None,
+        task_level: str = "graph",
+        return_type: str = "regression",
+        **kwargs,
+    ) -> Explanation:
+        """Compute the explanation.
+
+        Args:
+            x (torch.Tensor): node features.
+            edge_index (torch.Tensor): edge indices.
+            model (torch.nn.Module): the model to explain.
+            target (torch.Tensor): the target of the model.
+            target_index (int): the index of the target to explain.
+                By default suppose the target is a single value.
+            batch (torch.Tensor, optional): batch indicator.
+            **kwargs: additional arguments to pass to the GNN.
+        """
+
+    @abstractmethod
```

Comment:
Drop IMO.

Reply:
So I switched `Data` to  `**kwargs` but did not include `*args`, because when using `*args` you have to define an ordering of the input you expect, which may vary between two instances of `ExplainerAlgorithm` and make tests like this which I find messy: 
https://github.com/pyg-team/pytorch_geometric/blob/master/torch_geometric/nn/models/explainer.py#L81-L105.

Is this ok or are the `*args` necessary somehow ?